warmup.py

```diff
@@ -103,7 +105,18 @@ def energy(self, r):
             If ``r`` is 0, the energy is :py:obj:`numpy.inf`.
 
         """
-        pass
+        r = np.atleast_1d(r)
+        u = np.zeros_like(r)
+        for i, ri in enumerate(r):
```

Comment:
This is all great! As a last step, can we now try to get rid of this loop using numpy boolean selectors? For example:
```python
u[np.isclose(r,0)] = np.inf
```
will set all values of u to infinity where r is close to zero. You can do something similar to evaluate the potential in the range where `r <= rcut`. Because you initialized u with zeros, you do not need to set anything for `r > rcut` anymore. The force method should be updated similarly. Working on arrays (rather than entries in a loop) allows NumPy to do whatever calculation is optimal under the hood.

https://numpy.org/doc/stable/reference/arrays.indexing.html#boolean-array-indexing